### PR TITLE
Client/Server Connection and cleanup

### PR DESCRIPTION
**Pull request by pveentjer**
Connection had conflated concerns; it contained client and server side specific logic. Therefor 2 new interfaces that extend Connection were introduced:
1) ClientConnection
2) ServerConnection

The client specific logic is move to ClientConnection and server specific knowledge is moved to ServerConnection.

The ClientConnection (the actual implementation) is renamed to TcpClientConnection.
The ServerConnection (the actual implementation)  is renamed to TcpServerConnection

The ClientConnectionManagerImpl is renamed to TcpClientConnectionManager (which matches TcpServerConnectionManager).

com.hazelcast.client.impl.connection.nio has been renamed to com.hazelcast.client.impl.connection.tcp. That it is using nio isn't very relevant. The naming is now in line with what is being used in server.

The ServerConnection functionality has been split in an API part and an implementation part.
The API part can be found at com.hazelcast.internal.server
The TCP implementation part can be found at com.hazelcast.internal.server.tcp

EndpointManager is renamed to ServerConnectionManager because it ServerConnections are managed; not endpoints. Also the name is now in line with the ClientConnectionManager.

AggregateEndpointManager is renamed to AggregateServerConnectionManager

Connection.endpoint methods have been renamed to remoteAddress. 

NetworkingService is renamed to Server
And ServerNetworkingService is renamed to TcpServer

For EE see https://github.com/hazelcast/hazelcast-enterprise/pull/3575

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/15/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/15/Hazelcast-pr-builder-15.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 1 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/16/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/16/Hazelcast-pr-builder-16.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[44,24] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[49,8] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[67,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[69,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[71,70] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[73,23] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,30] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,45] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[98,44] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/HeartbeatManager.java:[76,96] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/ClientICMPManager.java:[110,104] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/statistics/ClusterConnectionMetricsProvider.java:[37,90] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[22,15] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[44,24] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[49,8] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[67,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[69,19] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[71,70] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[73,23] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,30] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[75,45] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/memory/impl/UnsafeUtil.java:[98,44] Unsafe is internal proprietary API and may be removed in a future release
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/HeartbeatManager.java:[76,96] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/ClientICMPManager.java:[110,104] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/statistics/ClusterConnectionMetricsProvider.java:[37,90] error: incompatible types: ClientConnection cannot be converted to TcpClientConnection
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/17/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/17/Hazelcast-pr-builder-17.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:check (default) on project hazelcast: failed with 1 bugs and 0 errors -> [Help 1]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/19/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/19/Hazelcast-pr-builder-19.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientSelectorManagementCenterClientTest.testManagementCenterClient_doesNotGetDisconnected:63->HazelcastTestSupport.assertTrueAllTheTime:1200->lambda$testManagementCenterClient_doesNotGetDisconnected$0:65 expected:<1> but was:<0>
[ERROR] Errors: 
[ERROR]   ClientAuthenticationTest.testAuthentication_with_mcModeEnabled_when_clusterStart_isNotComplete:111 ? IllegalState
[INFO] 
[ERROR] Tests run: 24273, Failures: 1, Errors: 1, Skipped: 995
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * Use this class to add client specific method.
+ */
+public interface ClientConnection extends Connection {
+
+    /**
+     * Returns the ClientConnectionManager of this ClientConnection.
+     *
+     * @return the ClientConnectionManager.
+     */
+    ClientConnectionManager getConnectionManager();
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java:39]: I could not find the usage of this method. Is this on enterprise side ? 
Do you have EE side ready for this pr ? Asking to check them together, to make sure this kind of changes.

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java [hunk 2]
```diff
@@ -534,6 +535,10 @@ private void shutdownWithExternalThread() {
 
     @Override
     public Collection<ClientConnection> getActiveConnections() {
+        return (Collection) activeConnections.values();
+    }
+
+    public Collection<TcpClientConnection> getActiveTcpClientConnections() {
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:541]: This second method is not necessary. The users of `getActiveTcpClientConnections` can use `getActiveConnections` instead. I have checked them. `ClientConnection` interface is enough for them. They don't need need implementation class.

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java (hunk 1 above)
Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java:39]: It seems it isn't used. I'll remove the method.

I'm working on the EE portion.

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java (hunk 2 above)
Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:541]: Fixed

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/32/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/32/Hazelcast-pr-builder-32.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllMessageTask.java:50: Line is longer than 130 characters (found 153). [LineLength]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/33/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/33/Hazelcast-pr-builder-33.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllMessageTask.java:50: Line is longer than 130 characters (found 153). [LineLength]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/34/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/34/Hazelcast-pr-builder-34.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllMessageTask.java:50: Line is longer than 130 characters (found 153). [LineLength]
--------------------------
</details>
</pre>

**Pull request merged**